### PR TITLE
Mephisto image-chat task

### PR DESCRIPTION
**Pull request by EricMichaelSmith**
**Patch description**
Add a task to have a human chat with a model about an image shown at the beginning of the conversation. The blueprint and world for this new task share the same base classes with those of the `turn_annotations` Mephisto task, which has now been renamed `model_chat` for generality.

Implementation details:
- The new image-chat task can be called with `python parlai/crowdsourcing/tasks/model_chat/run_image_chat.py`
  - A helper script, `parlai/crowdsourcing/tasks/model_chat/scripts/save_image_contexts.py`, has been added to create a file of images to choose among when collecting conversations.
  - A stack object, `ImageStack`, is used for keeping track of which images have been shown to which users when discussing them with which models.
  - For greater flexibility, ParlAI args for the models with which to run image chat are specified by passing in a JSON file with the `mephisto.blueprint.model_opt_path` arg.
- The annotations of the former `turn_annotations` task can be disabled by setting `mephisto.blueprint.annotations_config_path` to `""`

**Testing steps**
- `tests/crowdsourcing/tasks/model_chat/test_model_chat.py`
- `tests/crowdsourcing/tasks/model_chat/test_model_image_chat.py`
- `tests/crowdsourcing/tasks/model_chat/test_image_stack.py`

View of human/model image chat:
Screen Shot 2020-12-22 at 5 08 53 PM

### CODE REVIEW

**Comment by EricMichaelSmith** [comment]
@stephenroller @JackUrb @klshuster @mwillwork @jxmsML I'm marking this PR as ready for review now - the crowdsourcing CI check will not fully pass yet (see the PR description for an explanation), but the rest of it should be ready to go

File: parlai/crowdsourcing/tasks/model_chat/task_config/image_model_opts.yaml
Comment (klshuster) [inline parlai/crowdsourcing/tasks/model_chat/task_config/image_model_opts.yaml:1]: what is the `>` for? and is there a reason this isn't setup the normal yaml way? like `model_file: <mf>`

**Review by klshuster — APPROVED** [review]
cool! again, I would wait for a review from someone more familiar with mephisto/js, but i did have a bunch of nits/questions 😄

File: parlai/crowdsourcing/tasks/model_chat/README.md
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/model_chat/README.md:35]: Yes, that's fair - I've just changed the default val of this param to `""`. However, I've set this param to `${task_dir}/task_config/annotations_config.json` in `example.yaml`, because I think it'd be useful for users to see how annotation works

File: parlai/crowdsourcing/tasks/model_chat/README.md
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/model_chat/README.md:48]: Yes, changed

File: parlai/crowdsourcing/tasks/model_chat/README.md
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/model_chat/README.md:52]: Yes, good idea - I've added this as a line item to our `parlai.crowdsourcing` feature tracker

File: parlai/crowdsourcing/tasks/model_chat/task_config/image_model_opts.yaml
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/model_chat/task_config/image_model_opts.yaml:1]: Ah, the `>` allows you to specify a block of text. The reason I have the params set up like this is so I can parse the args using ParlaiParser, which I generally try to use as much as possible to avoid edge cases from differences in how params are parsed on the command line vs. when passing in an `Opt`. (It's related to the point you make above about some params having a `dist` specified, i.e. there are subtle differences in the flag depending on how it's used.) I also think it's useful to keep the normal command-line syntax here so that users can just specify commands exactly how they'd do it on the command line, to minimize confusion

File: parlai/crowdsourcing/tasks/model_chat/bot_agent.py
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/model_chat/bot_agent.py:56]: Yes, good point - changed

File: parlai/crowdsourcing/tasks/model_chat/bot_agent.py
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/model_chat/bot_agent.py:61]: Hmm, looks like that's a weird vestige - removing

File: parlai/crowdsourcing/tasks/model_chat/bot_agent.py
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/model_chat/bot_agent.py:68]: Yes - changed

File: parlai/crowdsourcing/tasks/model_chat/bot_agent.py
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/model_chat/bot_agent.py]: Changed

File: parlai/crowdsourcing/tasks/model_chat/bot_agent.py
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/model_chat/bot_agent.py:153]: Hmm, this logic is if the model file doesn't exist but is instead downloaded, like DialoGPT

File: parlai/crowdsourcing/tasks/model_chat/bot_agent.py
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/model_chat/bot_agent.py:171]: Good point, and also that's much cleaner - changed

File: parlai/crowdsourcing/tasks/model_chat/model_chat_blueprint.py
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/model_chat/model_chat_blueprint.py:200]: Good catch - added

File: parlai/crowdsourcing/tasks/model_chat/model_chat_blueprint.py
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/model_chat/model_chat_blueprint.py]: Yup, made controllable

File: parlai/crowdsourcing/tasks/model_chat/model_chat_blueprint.py
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/model_chat/model_chat_blueprint.py]: hmm, each value of the outer dict is the output of `Agent.share()`, which is itself a dict. I've tweaked the language a bit to hopefully clarify

File: parlai/crowdsourcing/tasks/model_chat/model_chat_blueprint.py
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/model_chat/model_chat_blueprint.py:370]: Ah, thanks for pointing that out - we only care about an onboarding data file if we're doing annotations, since the onboarding involves learning how to do annotations. I've added a comment to explain that

File: parlai/crowdsourcing/tasks/model_chat/model_chat_blueprint.py
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/model_chat/model_chat_blueprint.py:399]: Same as above - added a comment here as well

File: parlai/crowdsourcing/tasks/model_chat/model_chat_blueprint.py
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/model_chat/model_chat_blueprint.py:520]: No - currently, for turn annotations you specify the models by giving the path to a folder of models, and for image chat you specify the models with a YAML file. I've added a line item to the feature tracker to either support both methods universally or just one (but will solicit feedback on this before implementing). I've also added a note to the README to explain this current behavior

**Comment by EricMichaelSmith** [comment]
Thanks @klshuster for very useful comments! Just finished addressing them

File: parlai/crowdsourcing/tasks/model_chat/task_config/image_model_opts.yaml
Comment (klshuster) [reply parlai/crowdsourcing/tasks/model_chat/task_config/image_model_opts.yaml:1]: yeah that's fine with me

**Pull request merged**
